content/5_Plat/Matrix_Expo.mdx

```diff
@@ -129,6 +129,61 @@ int main() {
 }
 ```
 </CPPSection>
+<JavaSection>
+
+```java
+
+import java.util.*;
+import java.io.*;
+
+class FibonacciNumbers {
+	public final static int MOD = 1000000007;
+	public static long[][] m = {{1, 1}, {1, 0}};
+	public static long[][] m1 = {{1, 0}, {0, 1}};
+	public static long[][] mul(long[][] A, long[][] B, int MOD) {
+		long[][] C = {{0, 0}, {0, 0}};
+		for (int i = 0; i < 2; i++) {
+			for (int j = 0; j < 2; j++) {
+				for (int k = 0; k < 2; k++) {
+					C[i][j] += ((A[i][k]) * (B[k][j])) % MOD;
+				}
+			}
+		}
+    		return C;
+	}
+
+	public static long[][] power(long[][] matrix, long n, int MOD) {
+		if (n == 0) {
+			return m1;
+		}
+    		if (n == 1) {
+        		return matrix;
+    		}
+		long[][] half = power(matrix, n / 2, MOD);
+		long[][] square = mul(half, half, MOD);
+		if (n % 2 == 0) {
+			return square;
+		}
+		return mul(matrix, square, MOD);
+	}
+	
+	public static void main(String[] args) {
+		Scanner sc = new Scanner(System.in);
```

Comment:
don't use scanner

Reply:
Should I use buffer reader or something?

Reply:
yes